File: SimpleTunnelSSHForLinux/src/SimpleTunnelSSH.App/ConnectionEditorForm.cpp
#include "ConnectionEditorForm.h"

#include "AppIconFactory.h"

#include <QDialogButtonBox>
#include <QFileDialog>
#include <QFormLayout>
#include <QHBoxLayout>
#include <QLineEdit>
#include <QMessageBox>
#include <QPushButton>
#include <QSpinBox>
#include <QVBoxLayout>

namespace SimpleTunnelSSH::App
{

using SimpleTunnelSSH::Core::Models::SshConnectionProfile;

ConnectionEditorForm::ConnectionEditorForm(const SshConnectionProfile* connection, QWidget* parent)
    : QDialog(parent)
    , _nameEdit(new QLineEdit(this))
    , _hostEdit(new QLineEdit(this))
    , _portSpinBox(new QSpinBox(this))
    , _userEdit(new QLineEdit(this))
    , _identityFileEdit(new QLineEdit(this))
    , _result(connection == nullptr ? SshConnectionProfile {} : connection->deepClone())
{
    setWindowTitle(connection == nullptr ? QStringLiteral("Add connection") : QStringLiteral("Edit connection"));
    setModal(true);
    setMinimumWidth(520);
    setWindowFlag(Qt::WindowContextHelpButtonHint, false);

    _nameEdit->setText(_result.name);
    _hostEdit->setText(_result.host);
    _portSpinBox->setRange(1, 65535);
    _portSpinBox->setValue(_result.port <= 0 ? 22 : _result.port);
    _userEdit->setText(_result.userName);
    _identityFileEdit->setText(_result.identityFilePath);

    auto* browseButton = new QPushButton(QStringLiteral("Browse"), this);
    connect(browseButton, &QPushButton::clicked, this, &ConnectionEditorForm::browseIdentityFile);

    auto* identityLayout = new QHBoxLayout();
    identityLayout->setContentsMargins(0, 0, 0, 0);
    identityLayout->addWidget(_identityFileEdit);
    identityLayout->addWidget(browseButton);

    auto* formLayout = new QFormLayout();
    formLayout->addRow(QStringLiteral("Name"), _nameEdit);
    formLayout->addRow(QStringLiteral("Host"), _hostEdit);
    formLayout->addRow(QStringLiteral("Port"), _portSpinBox);
    formLayout->addRow(QStringLiteral("User"), _userEdit);
    formLayout->addRow(QStringLiteral("Identity file"), identityLayout);

    auto* buttonBox = new QDialogButtonBox(QDialogButtonBox::Save | QDialogButtonBox::Cancel, this);
    AppIconFactory::applyDialogButtonIcons(buttonBox);
    connect(buttonBox, &QDialogButtonBox::accepted, this, &ConnectionEditorForm::accept);
    connect(buttonBox, &QDialogButtonBox::rejected, this, &QDialog::reject);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(16, 16, 16, 16);
    layout->setSpacing(12);
    layout->addLayout(formLayout);
    layout->addWidget(buttonBox);
}

SshConnectionProfile ConnectionEditorForm::result() const
{
    return _result;
}

std::optional<SshConnectionProfile> ConnectionEditorForm::showDialog(QWidget* owner, const SshConnectionProfile* connection)
{
    ConnectionEditorForm form(connection, owner);

    if (form.exec() != QDialog::Accepted)
    {
        return std::nullopt;
    }

    return form.result();
}

void ConnectionEditorForm::browseIdentityFile()
{
    const auto filePath = QFileDialog::getOpenFileName(
        this,
        QStringLiteral("Choose identity file"),
        _identityFileEdit->text(),
        QStringLiteral("Private key files (*.pem *.ppk *.*)"));

    if (!filePath.isEmpty())
    {
        _identityFileEdit->setText(filePath);
    }
}

void ConnectionEditorForm::accept()
{
    if (_nameEdit->text().trimmed().isEmpty() || _hostEdit->text().trimmed().isEmpty() || _userEdit->text().trimmed().isEmpty())
    {
        QMessageBox::warning(this, QStringLiteral("Validation error"), QStringLiteral("Name, host and user are required."));
        return;
    }

    _result.name = _nameEdit->text().trimmed();
    _result.host = _hostEdit->text().trimmed();
    _result.port = _portSpinBox->value();
    _result.userName = _userEdit->text().trimmed();
    _result.identityFilePath = _identityFileEdit->text().trimmed();
    QDialog::accept();
}

} // namespace SimpleTunnelSSH::App

File: SimpleTunnelSSHForWindows/src/SimpleTunnelSSH.App/ConnectionEditorForm.cs
using SimpleTunnelSSH.Core.Models;

namespace SimpleTunnelSSH.App;

public sealed class ConnectionEditorForm : Form
{
    private readonly TextBox _nameTextBox;
    private readonly TextBox _hostTextBox;
    private readonly NumericUpDown _portNumericUpDown;
    private readonly TextBox _userTextBox;
    private readonly TextBox _identityFileTextBox;

    private ConnectionEditorForm(SshConnectionProfile? connection)
    {
        Text = connection is null ? "Add connection" : "Edit connection";
        StartPosition = FormStartPosition.CenterParent;
        FormBorderStyle = FormBorderStyle.FixedDialog;
        MaximizeBox = false;
        MinimizeBox = false;
        ShowInTaskbar = false;
        Width = 520;
        AutoSize = true;
        AutoSizeMode = AutoSizeMode.GrowAndShrink;

        _nameTextBox = new TextBox { Text = connection?.Name ?? string.Empty, Dock = DockStyle.Fill };
        _hostTextBox = new TextBox { Text = connection?.Host ?? string.Empty, Dock = DockStyle.Fill };
        _portNumericUpDown = new NumericUpDown { Minimum = 1, Maximum = 65535, Value = connection?.Port ?? 22, Dock = DockStyle.Fill };
        _userTextBox = new TextBox { Text = connection?.UserName ?? string.Empty, Dock = DockStyle.Fill };
        _identityFileTextBox = new TextBox { Text = connection?.IdentityFilePath ?? string.Empty, Dock = DockStyle.Fill };

        var browseButton = new Button { Text = "Browse", AutoSize = true };
        browseButton.Click += (_, _) => BrowseIdentityFile();

        var identityPanel = new TableLayoutPanel { Dock = DockStyle.Fill, ColumnCount = 2, AutoSize = true };
        identityPanel.ColumnStyles.Add(new ColumnStyle(SizeType.Percent, 100));
        identityPanel.ColumnStyles.Add(new ColumnStyle(SizeType.AutoSize));
        identityPanel.Controls.Add(_identityFileTextBox, 0, 0);
        identityPanel.Controls.Add(browseButton, 1, 0);

        var layout = new TableLayoutPanel
        {
            Dock = DockStyle.Fill,
            Padding = new Padding(16),
            ColumnCount = 2,
            AutoSize = true
        };
        layout.ColumnStyles.Add(new ColumnStyle(SizeType.AutoSize));
        layout.ColumnStyles.Add(new ColumnStyle(SizeType.Percent, 100));

        AddRow(layout, "Name", _nameTextBox, 0);
        AddRow(layout, "Host", _hostTextBox, 1);
        AddRow(layout, "Port", _portNumericUpDown, 2);
        AddRow(layout, "User", _userTextBox, 3);
        AddRow(layout, "Identity file", identityPanel, 4);

        var okButton = new Button { Text = "Save", DialogResult = DialogResult.OK, AutoSize = true };
        okButton.Click += (_, eventArgs) =>
        {
            if (!ValidateInput())
            {
                DialogResult = DialogResult.None;
            }
        };

        var cancelButton = new Button { Text = "Cancel", DialogResult = DialogResult.Cancel, AutoSize = true };
        var buttonsPanel = new FlowLayoutPanel
        {
            Dock = DockStyle.Bottom,
            FlowDirection = FlowDirection.RightToLeft,
            AutoSize = true,
            Padding = new Padding(16, 0, 16, 16)
        };
        buttonsPanel.Controls.Add(cancelButton);
        buttonsPanel.Controls.Add(okButton);

        Controls.Add(layout);
        Controls.Add(buttonsPanel);

        AcceptButton = okButton;
        CancelButton = cancelButton;
        AppTheme.Apply(this);
    }

    public static SshConnectionProfile? ShowDialog(IWin32Window owner, SshConnectionProfile? connection)
    {
        using var form = new ConnectionEditorForm(connection?.DeepClone());

        if (form.ShowDialog(owner) != DialogResult.OK)
        {
            return null;
        }

        var result = connection?.DeepClone() ?? new SshConnectionProfile();
        result.Name = form._nameTextBox.Text.Trim();
        result.Host = form._hostTextBox.Text.Trim();
        result.Port = (int)form._portNumericUpDown.Value;
        result.UserName = form._userTextBox.Text.Trim();
        result.IdentityFilePath = form._identityFileTextBox.Text.Trim();
        return result;
    }

    private void BrowseIdentityFile()
    {
        using var dialog = new OpenFileDialog
        {
            Filter = "Private key files (*.pem;*.ppk;*.*)|*.pem;*.ppk;*.*",
            RestoreDirectory = true,
            CheckFileExists = true
        };

        if (dialog.ShowDialog(this) == DialogResult.OK)
        {
            _identityFileTextBox.Text = dialog.FileName;
        }
    }

    private bool ValidateInput()
    {
        if (string.IsNullOrWhiteSpace(_nameTextBox.Text) || string.IsNullOrWhiteSpace(_hostTextBox.Text) || string.IsNullOrWhiteSpace(_userTextBox.Text))
        {
            MessageBox.Show(this, "Name, host and user are required.", "Validation error", MessageBoxButtons.OK, MessageBoxIcon.Warning);
            return false;
        }

        return true;
    }

    private static void AddRow(TableLayoutPanel layout, string labelText, Control control, int rowIndex)
    {
        layout.RowStyles.Add(new RowStyle(SizeType.AutoSize));
        layout.Controls.Add(new Label
        {
            Text = labelText,
            AutoSize = true,
            Padding = new Padding(0, 6, 12, 0)
        }, 0, rowIndex);
        layout.Controls.Add(control, 1, rowIndex);
    }
}